src/ring_theory/prime_power.lean

```diff
@@ -0,0 +1,311 @@
+/-
+Copyright (c) 2021 . All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anne Baanen, Paul Lezeau
+-/
+
+import algebra.squarefree
+import ring_theory.unique_factorization_domain
+
+/-!
+
+# Chains of divisors
+
+The results in this file show that in the monoid `associates M` of a unique_factorization_monoid
+`M`, an element `a` is an n-th prime power iff its set of divisors is a strictly increasing chain
+of length `n + 1`, aka we can find strictly increasing bijection bewteen `fin (n + 1)` and
+the set of factors of `a`
+
+We can then use this to show that given `a b : M`, if there is an monotone bijection
+between the sets of factors of `associates.mk a` and `associates.mk b` then the prime
+factorisations of `a` and `b` have the same shape.
+
+
+## Main results
+- `prime_pow_has_chain` : existence of chain for prime powers
+- `eq_prime_pow_of_has_chain` : elements that have a chain are prime powers
+- `multiplicity_le_of_monotone` : if there is a monotone bijection `d` between the set
+  of factors of `a : associates M` and the set of factors of `b : associates N` then
+  for any prime `p ∣ a`, `multiplicity p a ≤ multiplicity (d p) b`.
+
+## Todo
+- show that under the assumptions of `multiplicity_le_of_monotone`, `d p` is prime. Applying
+  `multiplicity_le_of_monotone` on `d.symm` then gives us `multiplicity p a = multiplicity (d p) b`
+
+-/
+
+
+section factorisations_same_shape
+
+noncomputable theory
+open_locale classical
+
+open unique_factorization_monoid multiplicity irreducible
+
+variables {M : Type*} [cancel_comm_monoid_with_zero M]
+
+lemma pow_prime_has_chain {p : associates M} (n : ℕ) (hn : n ≠ 0) (hp : prime p) :
+  ∃ c : fin (n + 1) → associates M,
```

Comment:
is it worth making a structure for these chains? carrying round all of these hypotheses does not seem scalable

Reply:
I'm not sure how useful it would be for later use, since I only use the chains to prove `image_prime_pow_dvd_of_monotone`., but I'd be happy to add it in a subsequent PR just in case someone else needs it.

Reply:
oh, I see, this is all just for the final proof? then probably fine ^^